src/persist-client/src/dyn_cfg.rs

```diff
@@ -0,0 +1,409 @@
+    ///
+    /// External users likely want [Config::shared] instead.
+    fn shared<'a>(config: &Config<Self>, vals: &'a ConfigSet) -> Option<&'a Arc<Self::Shared>>;
+
+    /// Converts this type to its type-erased enum equivalent.
+    fn to_val(val: &Self) -> ConfigVal;
+
+    /// Retrieves the current config value of this type from a value of its
+    /// corresponding sharable type.
+    ///
+    /// External users likely want [Config::get] or [Config::get_from_shared]
+    /// instead.
+    fn get(x: &Self::Shared) -> Self;
+
+    /// Updates the sharable value for a config of this type to the given value.
+    fn set(x: &Self::Shared, val: Self);
+}
+
+/// An set of [Config]s with values independent of other [ConfigSet]s (even if
+/// they contain the same configs).
+#[derive(Clone, Debug, Default)]
+pub struct ConfigSet {
+    configs: BTreeMap<String, ConfigEntry>,
+}
+
+impl ConfigSet {
+    /// Adds the given config to this set.
+    ///
+    /// An alias for [Config::register], but taking and returning `Self` to
+    /// allow for easy chaining.
+    pub fn add<T: ConfigType>(mut self, config: &Config<T>) -> Self {
+        let config = ConfigEntry {
+            name: config.name,
+            desc: config.desc,
+            default: T::to_val(&config.default),
+            val: T::to_val(&config.default),
+        };
+        if let Some(prev) = self.configs.insert(config.name.to_owned(), config) {
+            panic!("{} registered twice", prev.name);
+        }
+        self
+    }
+
+    /// Returns the configs currently registered to this set.
+    pub fn entries(&self) -> impl Iterator<Item = &ConfigEntry> {
+        self.configs.values()
+    }
+}
+
+/// An entry for a config in a [ConfigSet].
+#[derive(Clone, Debug)]
+pub struct ConfigEntry {
+    name: &'static str,
+    desc: &'static str,
+    // WIP: It's clunky that `default` is a `ConfigVal`, which is an updatable
+    // shared value, when it will never change. Worth the boilerplate of
+    // inventing a ConfigRef enum?
+    default: ConfigVal,
+    val: ConfigVal,
+}
+
+impl ConfigEntry {
+    /// The name of this config.
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+
+    /// The description of this config.
+    pub fn desc(&self) -> &'static str {
+        self.desc
+    }
+
+    /// The default value of this config.
+    ///
+    /// This value is never updated.
+    pub fn default(&self) -> &ConfigVal {
+        &self.default
+    }
+
+    /// The sharable value of this config in the set.
+    pub fn val(&self) -> &ConfigVal {
+        &self.val
+    }
+}
+
+/// A type-erased [ConfigType::Shared] for when set of different types are
+/// stored in a collection.
+#[derive(Clone, Debug)]
+pub enum ConfigVal {
+    /// A `bool` shared value.
+    Bool(Arc<AtomicBool>),
+    /// A `usize` shared value.
+    Usize(Arc<AtomicU64>),
+    /// A `String` shared value.
+    String(Arc<RwLock<String>>),
```

Comment:
Poked at it.

If we wanted to wrap the `BTreeMap`, it looks like the critical annoyance is `ConfigSet::add`, which is the only place that we require exclusive access to the map. Since we do all our adds at the beginning, this isn't a dealbreaker necessarily -- the fancy way to solve this would be to have a `ConfigSetBuilder`, I guess? Up to you whether that's worth the hassle.

Reply:
Tried this. Ended up running into the `Config::shared` method, which I realized was my original answer to this issue! (The hazards of only working on something on Fridays, I suppose.) That allows perf-sensitive users to both avoid this as well and do the lookup of the name in the BTreeMap only once. It does double the number of lines of code needed for a get from 1 to 2, but my thinking was that was reasonable for uses that opt in to being more perf-sensitive.

Thoughts? Also happy to remove that method and go with the `ConfigSetBuilder` if we think we'd rather optimize the common-case a bit in exchange for losing the ability to amortize the name lookup.

Reply:
Eh... I suspect it's rare that someone will take advantage of the two-step thing - hot enough to notice a map lookup, not hot enough to care about an atomic read, don't want to cache - but I don't mind punting and seeing how it gets picked up in practice!